finish lab05 jdbc library implementation

scyber opened the pull request.

Commit:
finish lab05 jdbc library implementation

stvort commented on lab05-spring-dao/src/main/java/ru/otus/domain/Book.java:
```diff
@@ -0,0 +1,14 @@
+    private Long id;
+    private String name;
+    private Long authorId;
+    private Long genreId;
```
Сейчас, данный класс это не доменный объект, а просто представление строки БД через Java-класс. Поэтому (в т.ч. с точки зрения ООП), будет лучше: 
```
private Author author;
private Genre genre;
```

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+    @Autowired
+    private IOService ioService;
+    @Autowired
+    private MessageSource messageSource;
```
Мы **многократно** договаривались на занятиях внедрять все через конструктор

stvort replied:
Локализация в данном приложении не требуется по заданию и попадает под пункт "без фанатизма") В явном виде он тут не прописан, но можно считать что есть в каждом ДЗ)

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+@ShellComponent
+public class BookShellService {
+
+    @Autowired
+    private BookDao bookDao;
+    @Autowired
+    private AuthorDao authorDao;
+    @Autowired
+    private GenreDao genreDao;
```
Дао лучше напрямую не использовать, а добавить сервисный слой

stvort replied:
Внедрять и использовать компоненты (и не только) лучше по интерфейсу

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+package ru.otus.services;
```
Команды шелл же, не относятся к сервисам. Они ближе к контроллерам

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+    @ShellMethod(value = "finish", key = {"f"})
+    public void finish(){
+        System.exit(0);
+    }
```
У шелл есть встроенная команда "exit"

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+        book.setGenreId(genreId);
+        Long bookId = bookDao.save(book);
+        String doneMessage = messageSource.getMessage("book.done", null, Locale.getDefault());
+        ioService.outputString(bookName + " " + doneMessage + " "+ bookId);
```
Логику в командах шелл лучше не хранить. 1-3 строки, чтобы делегировать обработку запроса и все

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+    }
+    @ShellMethod(value = "showAllGenres", key = {"show_genres"})
+    public void showAllGenres(){
+        Iterable<Genre> genres = genreDao.findAll();
```
Предлагаю использовать более высокоуровневые интерфейсы. Типа `List`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellService.java:
```diff
@@ -0,0 +1,120 @@
+    public void showAllBooks(){
+        bookDao.findAll().forEach(book -> {
+            System.out.println(this.bookConverter.convert(book));
+        });
```
Для каждой книги вызовется конвертер, что приведет к 2000 запросам для 100 книг. Это N+1 проблема. Мы обсуждали методы ее решения для Jdbc в начале первого занятия по ORM

stvort commented on lab05-spring-dao/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,6 @@
+DROP TABLE IF EXISTS BOOKS ;
+DROP TABLE IF EXISTS GENRES ;
+CREATE TABLE AUTHORS(ID BIGINT AUTO_INCREMENT PRIMARY KEY, AUTHOR_FIRST_NAME VARCHAR(255), AUTHOR_SECOND_NAME VARCHAR(255), AUTHOR_LAST_NAME VARCHAR(255));
+CREATE TABLE BOOKS(ID BIGINT AUTO_INCREMENT PRIMARY KEY, NAME VARCHAR(255), AUTHOR_ID VARCHAR (255), GENRE_ID VARCHAR(255));
```
1. Лучше обеспечить целостность данных на уровне схемы БД, с помощью внешних ключей
2. Почему идентификаторы автора и жанра это `VARCHAR (255)`?

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,88 @@
+    private static final String SQL_UPDATE_AUTHORS = "update authors set " +
+            "author_first_name =: author_first_name, author_second_name = :author_second_name, author_last_name= :author_last_name " +
+            "where id = :id";
+    private static final String SQL_DELETE = "delete from authors where id = :id";
```
Как я говорил на занятии, запросы лучше писать по месту их использования, не делая их константами

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.Optional;
+
+@Repository
+public class AuthorDao implements CommonRepository<Author> {
```
В задании было:
> НЕ делать AbstractDao.

Обобщенных интерфейсов тоже делать не нужно. В данном случае это ничего не дает, но ухудшит читаемость и свяжет нас при разработке. Что будет если появится сущность которую нужно будет только добавлять в БД или только читать оттуда?

Так же, лучше не использовать слова "репозиторий" и "дао" вместе. Либо одно, либо другое

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,88 @@
+    private static final String SQL_DELETE = "delete from authors where id = :id";
+
+
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
`final`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/GenreDao.java:
```diff
@@ -0,0 +1,76 @@
+
+@Repository
+public class GenreDao implements CommonRepository<Genre> {
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
`final`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/repository/CommonRepository.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.repository;
+
+public interface CommonRepository<T>{
+    public Long save(T domain);
+    public void delete(Long id);
+    public T findById(Long id);
+    public Iterable<T> findAll();
```
`public` для методов интерфейсов не пишут

stvort commented on lab05-spring-dao/src/test/java/ru/otus/BookDaoTest.java:
```diff
@@ -0,0 +1,41 @@
+    private static final Long GENRE_ID = 1L;
+    private static final Long AUTHOR_ID = 1L;
+
+    @Autowired
+    BookDao bookDao;
+    @Autowired
+    GenreDao genreDao;
+    @Autowired
+    AuthorDao authorDao;
```
`private`

stvort commented on lab05-spring-dao/src/test/java/ru/otus/LibraryApplicationTests.java:
```diff
@@ -0,0 +1,90 @@
+import ru.otus.services.IOService;
+
+@SpringBootTest
+class LibraryApplicationTests {
```
Один тестовый класс - один тестируемый. Все в одном классе не тестируем. Этот момент обсуждался на занятии по разбору ДЗ

stvort commented on lab05-spring-dao/src/test/resources/schema.sql:
```diff
@@ -0,0 +1,6 @@
+DROP TABLE IF EXISTS AUTHORS ;
```
Мы договаривались не класть схему БД в тестовые ресурсы. Только тестовые данные

Commit:
доработка по замечаниям
 - бины перенесены в конструктор
 - убрана локализация
 - вынесена обработка из shell
 - разнесены тесты
 - вместо id авторов и жанров в Book вынесены сущности
 - упрощен конверторы

Commit:
доработка по замечаниям
 - очистка импортов

Commit:
доработка по замечаниям
 - удаление констант

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellController.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.shell.standard.ShellMethod;
+
+@ShellComponent
+public class BookShellController {
```
Раз уж мы переименовали класс в контроллер, то почему он в пакете с сервисами?

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellController.java:
```diff
@@ -0,0 +1,57 @@
+@ShellComponent
+public class BookShellController {
+
+    private BookShellProcessor bookShellProcessor;
```
`final`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookShellProcessor.java:
```diff
@@ -0,0 +1,98 @@
+@Component
+public class BookShellProcessor {
+
+    private final BookDao bookDao;
+    private final AuthorDao authorDao;
+    private final GenreDao genreDao;
```
Для этого все еще нужно сделать сервисную прослойку (для каждого дао). Она может ничего не делать, просто прокидывать вызовы до дао, но при наличии управления транзакциями, оно будет осуществляться именно в данных прослойках. Так же, там можно собрать сущность из кусков, т.е. в метод сервиса передается набор аргументов, сервис собирает сущность и передает ее дао

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.Optional;
+
+@Repository
+public class AuthorDao {
```
Задача дао, скрыть источник данных за общим интерфейсом. Тут нет интерфейса, по которому можно было бы всегда подменить реализацию, работающую с другим источником

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,71 @@
+@Repository
+public class AuthorDao {
+
+    private final String sql_query_for_all = "select id, name from authors a ";
+    private final String sql_query_author_by_id = sql_query_for_all + " where id = :id";
+    private final String sql_insert_authors = "insert into authors ( name ) values ( :name)";
+    private final String sql_update_authors = "update authors set name = :name where id = :id";
+    private final String sql_delete = "delete from authors where id = :id";
```
> Как я говорил на занятии, запросы лучше писать по месту их использования, не делая их константами

Замечание не исправлено. Ну и в Java есть соглашение о наименовании. И по нему с подчеркиваниями именуются только константы, но большими буквами. Так же, в "правилах оформления" указано, что константой считается только `static final`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,71 @@
+
+    public Long save(Author domain) {
+        if(domain.getId() != null){
+            Author author = new Author();
```
У нас вроде уже есть автор? Зачем еще одного создавать?

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    RowMapper<Author> authorRowMapper = (ResultSet rs, int rowNum) -> {
+        return new Author(rs.getLong(1), rs.getString(2));
```
А если в запросе поменяют порядок полей? Лучше по имени доставать

stvort commented on lab05-spring-dao/src/test/java/ru/otus/AuthorTest.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.stream.Collectors;
+
+@SpringBootTest
+public class AuthorTest {
```
`@JdbcTest`

Ну и это не тест автора, а тест дао. А значит он должен по другому называться и лежать в том же пакете, что и тестируемый класс

scyber replied:
Здесь подключаю дао - а его из @JdbcTest не достать

stvort commented on lab05-spring-dao/src/test/java/ru/otus/AuthorTest.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Test
+    @DisplayName("Тест создания и сохранения нового автора")
+    @Order(1)
```
Тесты не должны зависеть от порядка выполнения и от результатов выполнения друг-друга

stvort commented on lab05-spring-dao/src/test/java/ru/otus/BookTest.java:
```diff
@@ -0,0 +1,89 @@
+    @Test
+    @DisplayName("Тестирование операций с книгами")
+    public void testBookOperation(){
+//        Author author = new Author(1L, "TestAuthor");
+//        Genre genre = new Genre(1L, "TestGenre");
+//        Book book = new Book(1L, "TestBook", author,genre);
+//        List<Book> books = new ArrayList<>();
+//        books.add(book);
+//        BookDao bookDaoMock = Mockito.mock(BookDao.class);
+//        IOService ioService = Mockito.mock(IOServiceStreams.class);
+//
+//        BookShellProcessor shellProcessor = new BookShellProcessor(bookDao,authorDao,genreDao,bookConverter,
+//                                                                  genreConverter,authorConverter,ioService );
```
Не используемый код, в репозитории не храним

Commit:
- Исправление по замечаниям

Commit:
- Исправление по тесту

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookService.java:
```diff
@@ -0,0 +1,90 @@
+    }
+    @Transactional
+    public long addBook(String title, long authorId, long genreId ){
+        Author author = authorDao.findById(authorId).get();
```
Одним из пунктов презентации занятия по разбору ДЗ было то, что не стоит отмахиваться от возмжостей `Optional`, вызывая у него `get`. Что тут д.б. если пусто? Исключение? Тогда `orElseThrhow`. Значение по умолчанию? Тогда `orElse/orElseGet`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/controllers/LibraryController.java:
```diff
@@ -0,0 +1,83 @@
+    private final BookService bookService;
+    private final IOService ioService;
+
+    public LibraryController(BookService bookShellProcessor, IOService ioService) {
```
Нету же больше процессора?

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookService.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.stream.Collectors;
+
+@Component
+public class BookService {
```
Сервису нужен интерфейс

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.Optional;
+
+@Repository
+public class AuthorDao {
```
Я уже немного устал это писать, но 
>Задача дао, скрыть источник данных за общим интерфейсом. Тут нет интерфейса, по которому можно было бы всегда подменить реализацию, работающую с другим источником

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDao.java:
```diff
@@ -0,0 +1,69 @@
+@Repository
+public class AuthorDao {
+
+    private final String sqlQueryForAll = "select id, name from authors a ";
+    private final String sqlQueryAuthorById = sqlQueryForAll + " where id = :id";
+    private final String sqlInsertAuthor = "insert into authors ( name ) values ( :name)";
+    private final String sqlUpdateAuthor = "update authors set name = :name where id = :id";
+    private final String sqlDelete = "delete from authors where id = :id";
```
А так же

>Как я говорил на занятии, запросы лучше писать по месту их использования, не делая их константами

и
>Замечание не исправлено. Ну и в Java есть соглашение о наименовании. И по нему с подчеркиваниями именуются только константы, но большими буквами. Так же, в "правилах оформления" указано, что константой считается только static final

stvort commented on lab05-spring-dao/src/test/java/ru/otus/services/BookServiceTest.java:
```diff
@@ -0,0 +1,78 @@
+@SpringBootTest
+class BookServiceTest {
+
+    @Autowired
+    private BookDao bookDao;
+    @Autowired
+    private AuthorDao authorDao;
+    @Autowired
+    private GenreDao genreDao;
+    @Autowired
+    private GenreConverter genreConverter;
+    @Autowired
+    private BookConverter bookConverter;
+    @Autowired
+    private AuthorConverter authorConverter;
```
Как мы договаривались на занятии, все зависимости тестируемых классов мы мокаем

scyber replied:
Здесь задал вопрос не увидел к сожалению ответа ни в slack, также не понимаю что именно тут замокать дао или только сервис, оставил не закрытым.

Commit:
- Исправлены дао добавлены интерфейсы
- Добавлен интерфейс над сервисом Библиотеки
- Вынесены sql запросы

Commit:
- Доработаны тесты сервиса
- Добавлены метод updateById

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,67 @@
+            throw new FindItemExecption("Could not find Author by ID " + id, e);
+        }
+    }
+    @Transactional
```
Мы транзакции предлагали вешать во вполне определенных местах. Тем не менее не совсем понятно, почему именно тут она есть, а больше нигде нет

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,67 @@
+        return this.namedParameterJdbcTemplate.query(sqlQueryForAll, authorRowMapper);
+    }
+
+    RowMapper<Author> authorRowMapper = (ResultSet rs, int rowNum) -> new Author(rs.getLong("id"), rs.getString("name"));
```
`private final` + поля располагаются наверху класса

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/BookDaoJdbc.java:
```diff
@@ -0,0 +1,113 @@
+                " left join authors a on b.author_id = a.id " +
+                " left join genres g on b.genre_id = g.id " +
+                " WHERE b.id = :id";
+        try {
+         Map<String,Object> mappedParameters = Collections.singletonMap("id",id);
+         return Optional.ofNullable(this.jdbcTemplate.queryForObject(sqlQueryFindById, mappedParameters, bookRowMapper));
+        } catch (EmptyResultDataAccessException e){
+            throw new FindItemExecption("Could not find book by id " + id, e);
+        }
```
Очень странное форматирование. Все же предлагаю использовать Ctrl+Alt+L в IDEA, в каждом классе, перед коммитом. А так же Ctrl+Alt+O

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/BookDaoJdbc.java:
```diff
@@ -0,0 +1,113 @@
+        genre.setName(rs.getString("genre_name"));
+        book.setGenre(genre);
+        return book;
+    };
```
Такая большая лямбда, что ее проще вложенным классом сделать

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+        this.authorConverter = authorConverter;
+    }
+
+    @Transactional
+    @Override
+    public List<String> getAllBooks(){
+      return  bookDao.findAll().stream().map(book -> this.bookConverter.convert(book)).collect(Collectors.toList());
+    }
```
Чтение не подразумевает изменений в БД. Так что лучше `@Transactional(readOnly = true)`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/services/BookService.java:
```diff
@@ -0,0 +1,17 @@
+    List<String> getAllBooks();
+    String getBookById(long id);
+    long addBook(String title, long authorId, long genreId );
+    void delBook(long id);
```
Три буквы экономить нет смысла, а читаемость такая экономия снижает

stvort commented on lab05-spring-dao/src/test/java/ru/otus/services/BookServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+
+    @BeforeEach
+    void prepareMocks(){
+        bookDao = Mockito.mock(BookDaoJdbc.class);
+        authorDao = Mockito.mock(AuthorDaoJdbc.class);
+        genreDao = Mockito.mock(GenreDaoJdbc.class);
+        bookService = new BookServiceImpl(bookDao,authorDao,genreDao,
+                bookConverter,genreConverter,authorConverter);
```
В тестах с контекстом **все, что является компонентом**, мы берем **из контекста**. Моки в контекст складываем с помощью аннотации `@MockBean`

stvort commented on lab05-spring-dao/src/test/java/ru/otus/dao/BookDaoTest.java:
```diff
@@ -0,0 +1,80 @@
+    @Autowired
+    private GenreDaoJdbc genreDao;
+    @Autowired
+    private AuthorDaoJdbc authorDao;
```
`BookDaoJdbc` не использует другие дао. Так что нет смысла их класть в контекст

Commit:
- Доработаны тесты сервиса
- Вынесен мапер в класс
- Убран ошибочный траназкшенал
- Переведены в RO get методы
- отформатирован код

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/AuthorDaoJdbc.java:
```diff
@@ -0,0 +1,71 @@
+public class AuthorDaoJdbc implements AuthorDao{
+
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    RowMapper<Author> authorRowMapper = (ResultSet rs, int rowNum) -> new Author(rs.getLong("id"), rs.getString("name"));
```
Как уже писалось ранее, данное поле, как и другие поля, лучше сделать `private final`

stvort commented on lab05-spring-dao/src/main/java/ru/otus/dao/GenreDaoJdbc.java:
```diff
@@ -0,0 +1,72 @@
+        return this.namedParameterJdbcTemplate.query(sqlSelectAll, genreRowMapper);
+    }
+
+    private RowMapper<Genre> genreRowMapper = (ResultSet rs, int numRow) -> new Genre(rs.getLong("id"), rs.getString("name"));
```
Это осталось внизу, плюс так же надо `final`

stvort commented on lab05-spring-dao/src/test/java/ru/otus/services/BookServiceTest.java:
```diff
@@ -0,0 +1,136 @@
+
+    @BeforeEach
+    void prepareMocks() {
+        bookService = new BookServiceImpl(bookDao, authorDao, genreDao,
+                bookConverter, genreConverter, authorConverter);
```
> В тестах с контекстом **все, что является компонентом**, мы берем из контекста. Моки в контекст складываем с помощью аннотации `@MockBean`

Еще раз, "**все, что является компонентом берем из контекста**". Мы не создаем сервисы, дао, конвертеры и прочие компоненты через `new`. Мы метаданными формируем контекст, а потом берем из него все, что нужно для тестирования. А зависимости мокаем. Все. Т.е. тут д.б. шесть полей с `@MockBean` и `@Autowired` над `BookService`. Сейчас же получается следующее:
- Поднимается приложение целиком, 
- В т.ч. в контекст кладутся моки
- Туда же кладется сервис книг (далее BS1), куда внедряется три нормальных конвертера и три мока дао
- В тесте создается **еще один сервис книг** (BS2), который никак с контекстом не связан и в него мы передаем три нормальных конвертера и три мока дао внедренные в тест из контекста
- И вот BS2 мы и тестируем по факту. Не то, что лежит в контексте (мы даже не можем быть уверены, что там есть BS1), а то что мы создали в тесте
